# 🧮 NumPy Complete Guide

**NumPy** (Numerical Python) is the fundamental package for numerical computing in Python. It provides powerful data structures, implementing multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions.

📘 **NumPy Documentation01**: [Link](https://numpy.org/doc/)

📘 **NumPy Documentation02**: [Link](https://www.w3schools.com/python/numpy/default.asp)

---

## 📑 Contents

1. What is NumPy?
3. NumPy Arrays vs Python Lists
4. Creating Arrays
5. Array Attributes
6. Indexing & Slicing
7. Array Manipulation
8. Mathematical Operations
9. Broadcasting
10. Universal Functions (ufuncs)
11. Aggregations & Statistics
12. Boolean Masking
13. File I/O with NumPy
14. Random Module
15. Useful Tips & Tricks
16. Summary


## What is NumPy?

NumPy is a powerful library for numerical operations in Python. It introduces `ndarray`, a fast and memory-efficient multi-dimensional array for vectorized operations. It is the foundation for libraries like Pandas, SciPy, and scikit-learn.

## NumPy Arrays vs Python Lists

- **Python Lists** are flexible but slower and consume more memory.
- **NumPy Arrays** are typed, support vectorized operations, and are faster.
```python

```

In [1]:

import numpy as np
arr = np.array([1, 2, 3])
print(arr + 2)  # Vectorized

[3 4 5]


## Some Functions



```python
import numpy as np
a = np.array([1, 2, 3])
b = np.array([[1, 2, 3], [4, 5, 6]])
```

| Category               | Function Name                    | Description                                    | Syntax / Usage                    | Example Output                         |
| ---------------------- | -------------------------------- | ---------------------------------------------- | --------------------------------- | -------------------------------------- |
| **Creation**           | `np.array()`                     | Creates a NumPy array                          | `np.array([1, 2, 3])`             | `[1 2 3]`                              |
|                        | `np.zeros()`                     | Creates an array filled with 0s                | `np.zeros((2, 3))`                | `[[0. 0. 0.], [0. 0. 0.]]`             |
|                        | `np.ones()`                      | Creates an array filled with 1s                | `np.ones((2, 3))`                 | `[[1. 1. 1.], [1. 1. 1.]]`             |
|                        | `np.full()`                      | Creates an array filled with a specified value | `np.full((2, 2), 7)`              | `[[7 7], [7 7]]`                       |
|                        | `np.eye()`                       | Identity matrix                                | `np.eye(3)`                       | `[[1. 0. 0.], [0. 1. 0.], [0. 0. 1.]]` |
|                        | `np.arange()`                    | Creates array from range                       | `np.arange(1, 10, 2)`             | `[1 3 5 7 9]`                          |
|                        | `np.linspace()`                  | Creates array with `n` equally spaced values   | `np.linspace(0, 1, 5)`            | `[0. 0.25 0.5 0.75 1.]`                |
|                        | `np.random.rand()`               | Random values in `[0, 1)`                      | `np.random.rand(2, 2)`            | `[[0.64 0.12], [0.75 0.38]]`           |
|                        | `np.random.randint()`            | Random integers in range                       | `np.random.randint(1, 10, (2,2))` | `[[7 2], [4 5]]`                       |
| **Inspection**         | `a.shape`                        | Shape of array                                 | `a.shape`                         | `(3,)`                                 |
|                        | `a.ndim`                         | Number of dimensions                           | `a.ndim`                          | `1`                                    |
|                        | `a.size`                         | Total number of elements                       | `a.size`                          | `3`                                    |
|                        | `a.dtype`                        | Data type                                      | `a.dtype`                         | `int64` or `int32`                     |
|                        | `a.itemsize`                     | Bytes per element                              | `a.itemsize`                      | `8`                                    |
|                        | `a.nbytes`                       | Total bytes                                    | `a.nbytes`                        | `24`                                   |
| **Reshaping**          | `a.reshape()`                    | Changes shape                                  | `a.reshape((3, 1))`               | `[[1], [2], [3]]`                      |
|                        | `a.flatten()`                    | Flattens to 1D                                 | `b.flatten()`                     | `[1 2 3 4 5 6]`                        |
|                        | `np.ravel()`                     | Similar to flatten, returns view when possible | `np.ravel(b)`                     | `[1 2 3 4 5 6]`                        |
|                        | `np.expand_dims()`               | Expands dimensions                             | `np.expand_dims(a, axis=0)`       | `[[1, 2, 3]]`                          |
|                        | `np.squeeze()`                   | Removes single-dimensional entries             | `np.squeeze([[1]])`               | `1`                                    |
| **Indexing & Slicing** | `a[1]`, `a[1:3]`                 | Access and slice elements                      | `a[1]`, `a[1:3]`                  | `2`, `[2, 3]`                          |
|                        | `a[::2]`                         | Step slicing                                   | `a[::2]`                          | `[1 3]`                                |
|                        | `a[::-1]`                        | Reverse array                                  | `a[::-1]`                         | `[3 2 1]`                              |
|                        | `np.where()`                     | Conditional index                              | `np.where(a > 1)`                 | `(array([1, 2]),)`                     |
|                        | `np.take()`                      | Extracts elements                              | `np.take(a, [0, 2])`              | `[1 3]`                                |
| **Sorting**            | `np.sort()`                      | Sorts array                                    | `np.sort([3, 1, 2])`              | `[1 2 3]`                              |
|                        | `np.argsort()`                   | Returns sort indices                           | `np.argsort([3, 1, 2])`           | `[1 2 0]`                              |
| **Copying**            | `np.copy()`                      | Copies array                                   | `c = np.copy(a)`                  | Same as `a`                            |
|                        | `a.view()`                       | Creates view (linked)                          | `v = a.view()`                    | Changes in `a` affect `v`              |
| **Aggregation**        | `np.sum()`, `np.mean()`          | Total or average values                        | `np.mean(a)`                      | `2.0`                                  |
|                        | `np.max()`, `np.min()`           | Max or min                                     | `np.max(a)`                       | `3`                                    |
|                        | `np.std()`, `np.var()`           | Standard deviation and variance                | `np.std(a)`                       | `0.816`                                |
| **Logic / Condition**  | `np.any()`, `np.all()`           | Checks if any/all elements are True            | `np.any(a > 1)`                   | `True`                                 |
|                        | `np.equal()`, `np.array_equal()` | Element-wise comparison                        | `np.equal(a, [1, 0, 3])`          | `[True False True]`                    |
| **Linear Algebra**     | `np.dot()`, `np.matmul()`        | Dot product, matrix multiplication             | `np.dot(a, [4,5,6])`              | `32`                                   |
|                        | `np.linalg.inv()`                | Inverse of a matrix                            | `np.linalg.inv([[1,2],[3,4]])`    | Returns inverse                        |
|                        | `np.linalg.det()`                | Determinant                                    | `np.linalg.det([[1,2],[3,4]])`    | `-2.0`                                 |
|                        | `np.linalg.eig()`                | Eigenvalues and eigenvectors                   | `np.linalg.eig()`                 | Varies                                 |
| **Saving / Loading**   | `np.save()`, `np.load()`         | Save/load `.npy` files                         | `np.save('file.npy', a)`          | Binary file                            |
|                        | `np.savetxt()`, `np.loadtxt()`   | Save/load text (CSV, etc.)                     | `np.savetxt('file.txt', a)`       | Plain text                             |


In [6]:

np.zeros((2, 3))
np.ones((3,2))
np.arange(0, 10, 2)
np.linspace(0, 1, 5)
np.eye(4)

print(np.eye(4))

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


## Array Attributes

Check the structure of a NumPy array:

| Attribute      | Description                                                         | Example Usage    | Output / Explanation                         |
| -------------- | ------------------------------------------------------------------- | ---------------- | -------------------------------------------- |
| `arr.ndim`     | Returns the number of dimensions (axes)                             | `arr.ndim`       | `2` – it's a 2D array                        |
| `arr.shape`    | Tuple showing the size of each dimension                            | `arr.shape`      | `(2, 3)` – 2 rows, 3 columns                 |
| `arr.size`     | Total number of elements in the array                               | `arr.size`       | `6` – 2 × 3                                  |
| `arr.dtype`    | Data type of the array elements                                     | `arr.dtype`      | `dtype('int64')` (on most systems)           |
| `arr.itemsize` | Size (in bytes) of each element in the array                        | `arr.itemsize`   | `8` – each `int64` takes 8 bytes             |
| `arr.nbytes`   | Total bytes consumed by the array                                   | `arr.nbytes`     | `48` – 6 elements × 8 bytes                  |
| `arr.T`        | Transpose of the array (swaps axes)                                 | `arr.T`          | `[[1 4] [2 5] [3 6]]` – shape becomes (3,2)  |
| `arr.flat`     | 1D iterator over array elements                                     | `list(arr.flat)` | `[1, 2, 3, 4, 5, 6]`                         |
| `arr.real`     | Returns the real part (useful if array has complex numbers)         | `arr.real`       | Same as `arr` if real numbers                |
| `arr.imag`     | Returns the imaginary part (0 if no imaginary part)                 | `arr.imag`       | All zeros for real-valued arrays             |
| `arr.data`     | Buffer containing the actual array elements                         | `arr.data`       | Returns a memory buffer object               |
| `arr.base`     | References the original object (if view), else `None`               | `arr.base`       | `None` – this is the original array          |
| `arr.strides`  | Number of bytes to step in each dimension when traversing the array | `arr.strides`    | `(24, 8)` – row stride = 3×8, col stride = 8 |
| `arr.flags`    | Info about memory layout (C-contiguous, F-contiguous, etc.)         | `arr.flags`      | Shows memory layout flags                    |


In [9]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
print(arr.shape)   # (2, 3)
print(arr.ndim)    # 2
print(arr.dtype)   # int64

(2, 3)
2
int32


## Indexing & Slicing

Access elements or subarrays:


In [10]:
a = np.array([10, 20, 30, 40])
print(a[1])      # 20
print(a[1:3])    # [20 30]

20
[20 30]


## Array Manipulation


In [11]:
a = np.array([[1, 2], [3, 4]])
a.reshape((4,))
a.flatten()
a.T  # Transpose

array([[1, 3],
       [2, 4]])

## Mathematical Operations
Let’s assume:

```python
import numpy as np
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
```

| Operation                 | Description                                     | Syntax / Function             | Example Output                 |
| ------------------------- | ----------------------------------------------- | ----------------------------- | ------------------------------ |
| **Addition**              | Element-wise addition                           | `a + b` or `np.add(a, b)`     | `[5 7 9]`                      |
| **Subtraction**           | Element-wise subtraction                        | `a - b` or `np.subtract(a,b)` | `[-3 -3 -3]`                   |
| **Multiplication**        | Element-wise multiplication                     | `a * b` or `np.multiply(a,b)` | `[4 10 18]`                    |
| **Division**              | Element-wise division                           | `a / b` or `np.divide(a,b)`   | `[0.25 0.4 0.5]`               |
| **Power**                 | Element-wise exponentiation                     | `a ** 2` or `np.power(a, 2)`  | `[1 4 9]`                      |
| **Modulus**               | Element-wise modulus (remainder)                | `a % 2` or `np.mod(a, 2)`     | `[1 0 1]`                      |
| **Floor Division**        | Element-wise floor division (discard remainder) | `a // 2`                      | `[0 1 1]`                      |
| **Absolute**              | Element-wise absolute value                     | `np.abs([-1, -2, 3])`         | `[1 2 3]`                      |
| **Square Root**           | Element-wise square root                        | `np.sqrt(a)`                  | `[1. 1.41 1.73]`               |
| **Exponential**           | Element-wise exponential (e^x)                  | `np.exp(a)`                   | `[2.71 7.39 20.08]`            |
| **Logarithm (natural)**   | Element-wise natural log                        | `np.log(a)`                   | `[0. 0.69 1.09]`               |
| **Logarithm (base 10)**   | Logarithm base 10 of array elements             | `np.log10(a)`                 | `[0. 0.301 0.477]`             |
| **Max / Min**             | Maximum or minimum element                      | `np.max(a)` / `np.min(a)`     | `3` / `1`                      |
| **Sum**                   | Sum of all elements                             | `np.sum(a)`                   | `6`                            |
| **Mean**                  | Arithmetic mean                                 | `np.mean(a)`                  | `2.0`                          |
| **Median**                | Middle value                                    | `np.median(a)`                | `2.0`                          |
| **Standard Deviation**    | Spread of elements                              | `np.std(a)`                   | `0.816`                        |
| **Variance**              | Measure of variation                            | `np.var(a)`                   | `0.666`                        |
| **Dot Product**           | Inner product of vectors                        | `np.dot(a, b)`                | `32`                           |
| **Matrix Multiplication** | Matrix multiplication (2D+)                     | `np.matmul(a, b)`             | Only for 2D or reshaped arrays |
| **Cumulative Sum**        | Running total                                   | `np.cumsum(a)`                | `[1 3 6]`                      |
| **Cumulative Product**    | Running product                                 | `np.cumprod(a)`               | `[1 2 6]`                      |
| **Clip**                  | Limit values to within min and max              | `np.clip(a, 1, 2)`            | `[1 2 2]`                      |
| **Round**                 | Rounds elements to given decimals               | `np.round(np.sqrt(a), 2)`     | `[1. 1.41 1.73]`               |


📘 **NumPy Math Functions**: [Link](https://numpy.org/doc/stable/reference/routines.math.html)

In [20]:
# EXAMPLES
 
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
print(a + b)  # [5 7 9]
print(np.dot(a, b))  # 32
print(np.matmul(a,b))

[5 7 9]
32
32


## Broadcasting

Broadcasting is a trick NumPy uses to let you do math on arrays that aren't exactly the same size. NumPy attempts to make them compatible by applying a set of rules. 

Broadcasting Rules:

For two arrays to be broadcastable, NumPy compares their shapes element-wise, starting from the trailing (rightmost) dimensions:

    - Equal Dimensions: If the dimensions are equal, they are compatible.
    - One Dimension is 1: If one of the dimensions is 1, it is stretched to match the other dimension.
    - Missing Dimensions: If one array has fewer dimensions than the other, its shape is effectively "padded" with ones on its left side until both arrays have the same number of dimensions.
    - Incompatible: If none of the above conditions are met for any pair of dimensions, the arrays are not broadcastable, and NumPy will raise a ValueError.

In [21]:
a = np.array([[1], [2], [3]])
b = np.array([10, 20, 30])
print(a + b)

[[11 21 31]
 [12 22 32]
 [13 23 33]]


## Universal Functions (ufuncs)

Fast, element-wise functions:


In [23]:
np.sqrt([1, 4, 9])
np.exp([0, 1])
np.log([1, np.e])

array([0., 1.])

## Aggregations & Statistics



In [28]:

arr = np.array([[1, 2], [3, 4]])
print(arr.sum())
print(arr.mean(axis=0))
print(arr.max())
print(arr.std())


10
[2. 3.]
4
1.118033988749895


## Boolean Masking


In [29]:
arr = np.array([1, 2, 3, 4, 5])
mask = arr > 3
print(arr[mask])  # [4 5]

[4 5]


## File I/O with NumPy


In [31]:
arr = np.array([[1, 2, 3], [4, 5, 6]])
np.savetxt('data.txt', arr)                 # save file
loaded = np.loadtxt('data.txt')             # load file
print(loaded)

[[1. 2. 3.]
 [4. 5. 6.]]


## Random Module


| Function                                        | Description                                                       | Syntax / Usage Example                        | Example Output                             |
| ----------------------------------------------- | ----------------------------------------------------------------- | --------------------------------------------- | ------------------------------------------ |
| **Random Sampling**                             |                                                                   |                                               |                                            |
| `np.random.rand(d0, d1, ...)`                   | Uniform distribution over \[0, 1), returns samples in given shape | `np.random.rand(2, 2)`                        | `[[0.32 0.85], [0.76 0.19]]`               |
| `np.random.randn(d0, d1, ...)`                  | Standard normal distribution (mean=0, std=1)                      | `np.random.randn(3)`                          | `[ 0.24 -1.12  1.45]`                      |
| `np.random.randint(low, high, size)`            | Random integers in range \[low, high)                             | `np.random.randint(1, 10, size=(2,2))`        | `[[3 5], [8 1]]`                           |
| `np.random.random(size)`                        | Alias for `rand()`, samples from \[0, 1)                          | `np.random.random((2,))`                      | `[0.5, 0.9]`                               |
| `np.random.choice(a, size, replace, p)`         | Random samples from 1D array (with/without replacement)           | `np.random.choice([10, 20, 30], size=2)`      | `[30 10]`                                  |
| `np.random.bytes(length)`                       | Returns random bytes                                              | `np.random.bytes(5)`                          | `b'\xaf\x93\x84&\x12'`                     |
| **Distributions**                               |                                                                   |                                               |                                            |
| `np.random.uniform(low, high, size)`            | Uniform distribution \[low, high)                                 | `np.random.uniform(1.0, 2.0, 3)`              | `[1.2 1.7 1.9]`                            |
| `np.random.normal(loc, scale, size)`            | Normal (Gaussian) distribution                                    | `np.random.normal(0, 1, 4)`                   | `[-1.0, 0.6, 0.1, 0.9]`                    |
| `np.random.poisson(lam, size)`                  | Poisson distribution                                              | `np.random.poisson(3.0, 5)`                   | `[2 1 4 5 3]`                              |
| `np.random.binomial(n, p, size)`                | Binomial distribution                                             | `np.random.binomial(10, 0.5, 5)`              | `[4 6 5 7 3]`                              |
| `np.random.beta(a, b, size)`                    | Beta distribution                                                 | `np.random.beta(0.5, 0.5, 3)`                 | `[0.3 0.8 0.5]`                            |
| `np.random.gamma(shape, scale, size)`           | Gamma distribution                                                | `np.random.gamma(2.0, 1.5, 3)`                | `[2.1 3.0 0.9]`                            |
| `np.random.exponential(scale, size)`            | Exponential distribution                                          | `np.random.exponential(1.0, 5)`               | `[0.5 1.3 0.1 2.3 0.9]`                    |
| `np.random.geometric(p, size)`                  | Geometric distribution                                            | `np.random.geometric(0.5, 5)`                 | `[2 1 1 3 1]`                              |
| `np.random.lognormal(mean, sigma, size)`        | Log-normal distribution                                           | `np.random.lognormal(0.0, 1.0, 3)`            | `[1.2 2.7 0.8]`                            |
| `np.random.laplace(loc, scale, size)`           | Laplace distribution                                              | `np.random.laplace(0.0, 1.0, 3)`              | `[-0.5, 0.1, 1.3]`                         |
| `np.random.f(dfnum, dfden, size)`               | F (variance ratio) distribution                                   | `np.random.f(5, 2, 3)`                        | `[1.3 0.7 3.2]`                            |
| `np.random.chisquare(df, size)`                 | Chi-square distribution                                           | `np.random.chisquare(2, 4)`                   | `[0.9 1.1 2.4 3.2]`                        |
| `np.random.triangular(left, mode, right, size)` | Triangular distribution                                           | `np.random.triangular(1, 2, 3, 5)`            | `[2.1 2.3 1.8 1.9 2.0]`                    |
| `np.random.wald(mean, scale, size)`             | Wald (inverse Gaussian) distribution                              | `np.random.wald(3, 2, 4)`                     | `[3.2 2.9 3.5 2.7]`                        |
| `np.random.weibull(a, size)`                    | Weibull distribution                                              | `np.random.weibull(2.0, 5)`                   | `[0.4 0.7 1.1 1.4 1.2]`                    |
| `np.random.standard_normal(size)`               | Standard normal (mean=0, std=1)                                   | `np.random.standard_normal(3)`                | `[0.2, -0.8, 1.4]`                         |
| `np.random.standard_gamma(shape, size)`         | Standard Gamma (scale = 1)                                        | `np.random.standard_gamma(2.0, 4)`            | `[1.3 0.9 2.2 1.7]`                        |
| `np.random.standard_exponential(size)`          | Standard Exponential (scale = 1)                                  | `np.random.standard_exponential(3)`           | `[0.6 1.2 0.3]`                            |
| `np.random.standard_t(df, size)`                | Standard Student's t-distribution                                 | `np.random.standard_t(5, 4)`                  | `[0.4 1.1 -0.9 2.3]`                       |
| **Shuffling / Permutation**                     |                                                                   |                                               |                                            |
| `np.random.shuffle(x)`                          | Shuffles array in-place (only 1D or rows of 2D)                   | `x = np.array([1,2,3]); np.random.shuffle(x)` | `x` is now shuffled                        |
| `np.random.permutation(x)`                      | Returns a shuffled copy                                           | `np.random.permutation(5)`                    | `[4 2 1 3 0]`                              |
| **Seed Control**                                |                                                                   |                                               |                                            |
| `np.random.seed(seed)`                          | Sets the seed for reproducibility                                 | `np.random.seed(42)`                          | Ensures same output every run              |
| `np.random.get_state()`                         | Gets RNG state                                                    | `state = np.random.get_state()`               | Internal state (used to restore RNG state) |
| `np.random.set_state(state)`                    | Restores RNG state                                                | `np.random.set_state(state)`                  |                                            |


In [32]:
np.random.seed(42)
np.random.rand(2, 2)
np.random.randint(0, 10, size=(3,))

array([6, 9, 2])

## Useful Tips & Tricks

- Use `np.where()` for conditional logic
- Use `np.unique()` for unique values
- Use `np.concatenate()` to join arrays

In [33]:
np.where(arr > 2, 'Yes', 'No')

array([['No', 'No', 'Yes'],
       ['Yes', 'Yes', 'Yes']], dtype='<U3')

## Value changing example

In [2]:
import numpy as np
arr = np.array([[56,18,60,22,52],[12,11,81,21,2],[52,26,28,9,23],[61,29,67,22,77]])
arr

array([[56, 18, 60, 22, 52],
       [12, 11, 81, 21,  2],
       [52, 26, 28,  9, 23],
       [61, 29, 67, 22, 77]])

In [3]:
arr >50

array([[ True, False,  True, False,  True],
       [False, False,  True, False, False],
       [ True, False, False, False, False],
       [ True, False,  True, False,  True]])

In [ ]:
arr[arr > 50]

array([56, 60, 52, 81, 52, 61, 67, 77])

In [9]:
arr[(arr > 50) & (arr%2!=0) ] = 0
arr

array([[56, 18, 60, 22, 52],
       [12, 11,  0, 21,  2],
       [52, 26, 28,  9, 23],
       [ 0, 29,  0, 22,  0]])

## Summary

NumPy is essential for scientific computing and machine learning. Its fast, vectorized operations, broadcasting features, and extensive mathematical toolkit make it an indispensable tool in the Python ecosystem.
